File: src/bosh-nats-sync/lib/nats_sync/nats_auth_config.rb
module NATSSync
  class NatsAuthConfig
    def initialize(vms, director_subject, hm_subject)
      @vms = vms
      @hm_subject = hm_subject
      @director_subject = director_subject
      @config = { 'authorization' =>
                    { 'users' => [] } }
    end

    def director_user
      {
        'user' => @director_subject,
        'permissions' => {
          'publish' => %w[agent.* agent.inbox.> hm.director.alert],
          'subscribe' => ['director.>'],
        },
      }
    end

    def hm_user
      {
        'user' => @hm_subject,
        'permissions' => {
          'publish' => [],
          'subscribe' => %w[hm.agent.heartbeat.* hm.agent.alert.* hm.agent.shutdown.* hm.director.alert],
        },
      }
    end

    def agent_user(agent_id, cn)
      {
        'user' => "C=USA, O=Cloud Foundry, CN=#{cn}.agent.bosh-internal",
        'permissions' => {
          'publish' => [
            "hm.agent.heartbeat.#{agent_id}",
            "hm.agent.alert.#{agent_id}",
            "hm.agent.shutdown.#{agent_id}",
            "director.*.#{agent_id}.*",
            "director.agent.disk.*.#{agent_id}",
          ],
          "subscribe": ["agent.#{agent_id}", "agent.inbox.#{agent_id}.>"],
        },
      }
    end

    def create_config
      @config['authorization']['users'] << director_user unless @director_subject.nil?
      @config['authorization']['users'] << hm_user unless @hm_subject.nil?
      @vms.each do |vm|
        agent_id = vm['agent_id']
        if !vm['permanent_nats_credentials']
          @config['authorization']['users'] << agent_user(agent_id, agent_id + '.bootstrap')
        end
        @config['authorization']['users'] << agent_user(agent_id, agent_id)
      end
      @config
    end
  end
end

File: src/bosh-nats-sync/pkg/natsauthconfig/nats_auth_config.go
package natsauthconfig

import "fmt"

type VM struct {
	AgentID                 string `json:"agent_id"`
	PermanentNATSCredentials bool   `json:"permanent_nats_credentials"`
}

type Permissions struct {
	Publish   []string `json:"publish"`
	Subscribe []string `json:"subscribe"`
}

type User struct {
	User        string      `json:"user"`
	Permissions Permissions `json:"permissions"`
}

type Authorization struct {
	Users []User `json:"users"`
}

type AuthorizationConfig struct {
	Authorization Authorization `json:"authorization"`
}

func directorUser(subject string) User {
	return User{
		User: subject,
		Permissions: Permissions{
			Publish:   []string{"agent.*", "hm.director.alert"},
			Subscribe: []string{"director.>"},
		},
	}
}

func hmUser(subject string) User {
	return User{
		User: subject,
		Permissions: Permissions{
			Publish:   []string{},
			Subscribe: []string{"hm.agent.heartbeat.*", "hm.agent.alert.*", "hm.agent.shutdown.*", "hm.director.alert"},
		},
	}
}

func agentUser(agentID, cn string) User {
	return User{
		User: fmt.Sprintf("C=USA, O=Cloud Foundry, CN=%s.agent.bosh-internal", cn),
		Permissions: Permissions{
			Publish: []string{
				fmt.Sprintf("hm.agent.heartbeat.%s", agentID),
				fmt.Sprintf("hm.agent.alert.%s", agentID),
				fmt.Sprintf("hm.agent.shutdown.%s", agentID),
				fmt.Sprintf("director.*.%s.*", agentID),
			},
			Subscribe: []string{fmt.Sprintf("agent.%s", agentID)},
		},
	}
}

func CreateConfig(vms []VM, directorSubject, hmSubject *string) AuthorizationConfig {
	cfg := AuthorizationConfig{
		Authorization: Authorization{
			Users: []User{},
		},
	}

	if directorSubject != nil {
		cfg.Authorization.Users = append(cfg.Authorization.Users, directorUser(*directorSubject))
	}
	if hmSubject != nil {
		cfg.Authorization.Users = append(cfg.Authorization.Users, hmUser(*hmSubject))
	}

	for _, vm := range vms {
		if !vm.PermanentNATSCredentials {
			cfg.Authorization.Users = append(cfg.Authorization.Users, agentUser(vm.AgentID, vm.AgentID+".bootstrap"))
		}
		cfg.Authorization.Users = append(cfg.Authorization.Users, agentUser(vm.AgentID, vm.AgentID))
	}

	return cfg
}
